pkg/detectors/snowflake/snowflake.go

```diff
@@ -0,0 +1,161 @@
+}
+
+// Ensure the Scanner satisfies the interface at compile time.
+var _ detectors.Detector = (*Scanner)(nil)
+
+var (
+	accountIdentifierPat = regexp.MustCompile(detectors.PrefixRegex([]string{"account"}) + `\b([a-zA-Z]{7}-[0-9a-zA-Z]{7})\b`)
+	usernameExclusionPat = `!@#$%^&*{}:<>,.;?()/\+=\s\n`
+)
+
+const (
+	database                  = "SNOWFLAKE"
+	retrieveAllDatabasesQuery = "SHOW DATABASES"
+)
+
+// Keywords are used for efficiently pre-filtering chunks.
+// Use identifiers in the secret preferably, or the provider name.
+func (s Scanner) Keywords() []string {
+	return []string{"snowflake"}
+}
+
+func meetsSnowflakePasswordRequirements(password string) (string, bool) {
+	var hasLower, hasUpper, hasNumber, minLen bool
+
+	if len(password) < 8 {
+		minLen = false
+	} else {
+		minLen = true
+	}
+
+	for _, char := range password {
+		switch {
+		case unicode.IsLower(char):
+			hasLower = true
+		case unicode.IsUpper(char):
+			hasUpper = true
+		case unicode.IsNumber(char):
+			hasNumber = true
+		}
+
+		if hasLower && hasUpper && hasNumber && minLen {
+			return password, true
+		}
+	}
+
+	return "", false
+}
+
+// FromData will find and optionally verify Snowflake secrets in a given set of bytes.
+func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (results []detectors.Result, err error) {
+	dataStr := string(data)
+
+	accountMatches := accountIdentifierPat.FindAllStringSubmatch(dataStr, -1)
+
+	usernameRegexState := common.UsernameRegexCheck(usernameExclusionPat)
+	usernameMatches := usernameRegexState.Matches(data)
+
+	passwordRegexState := common.PasswordRegexCheck(" ") // No explicit character exclusions by Snowflake for passwords
+	passwordMatches := passwordRegexState.Matches(data)
+
+	for _, accountMatch := range accountMatches {
+		if len(accountMatch) != 2 {
+			continue
+		}
+		resAccountMatch := strings.TrimSpace(accountMatch[1])
+
+		for _, resUsernameMatch := range usernameMatches {
+
+			for _, resPasswordMatch := range passwordMatches {
+				_, metPasswordRequirements := meetsSnowflakePasswordRequirements(resPasswordMatch)
+
+				if !metPasswordRequirements {
+					continue
+				}
+
+				uri := fmt.Sprintf("%s:%s@%s/%s", resUsernameMatch, resPasswordMatch, resAccountMatch, database)
+
+				s1 := detectors.Result{
+					DetectorType: detectorspb.DetectorType_Snowflake,
+					Raw:          []byte(resPasswordMatch),
+					ExtraData: map[string]string{
+						"account":  resAccountMatch,
+						"username": resUsernameMatch,
+					},
+				}
+
+				if verify {
+					_ = &gosnowflake.Config{
+						Account:  resAccountMatch,
+						User:     resUsernameMatch,
+						Password: resPasswordMatch,
+						Database: database,
+					}
+
+					// Open a connection to Snowflake
+					db, err := sql.Open("snowflake", uri) // Needs the snowflake driver from gosnowflake
+
+					if err != nil {
+						s1.VerificationError = fmt.Errorf("unable to open a connection to Snowflake %+v", err)
+					}
+					defer db.Close()
```

Comment:
Question: Is it okay to Close this if there was an error? With net/http requests the request would be nil and have a nil pointer exception

Reply:
when i tried testing it with bad credentials, it errors without any NPE/panic

Reply:
if i malform the connection string like 
> uri := fmt.Sprintf("%s://^#%s@%s/%s", resUsernameMatch, resPasswordMatch, resAccountMatch, database)

it will still generate non-nil `db`

example:

```
db:  &{{{} {} 0} {zubairkhan14://^#fakepassword1!@tuacoip-zt74995/SNOWFLAKE 0x104642948} {{} {} 0} {0 0} [] map[] 0 0 0x140001f07e0 false map[] map[] 0 0 0 0 <nil> 0 0 0 0 0x10278a0f0}
```

Reply:
the only condition in which sql.Open returns an empty response with error is if the driver does not exist - which we are hard setting to snowflake

```
func Open(driverName, dataSourceName string) (*DB, error) {
	driversMu.RLock()
	driveri, ok := drivers[driverName]
	driversMu.RUnlock()
	if !ok {
		return nil, fmt.Errorf("sql: unknown driver %q (forgotten import?)", driverName)
	}
```